File: edge/forwarder/src/forwarder/streams.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from lxml import etree

CATEGORY_PARENTS: dict[str, Literal["SAMPLE", "EVENT", "CONDITION"]] = {
    "Samples": "SAMPLE",
    "Events": "EVENT",
    "Condition": "CONDITION",
}
CONDITION_LEVELS = {"Normal", "Warning", "Fault", "Unavailable"}


@dataclass(frozen=True)
class Observation:
    device_uuid: str
    sequence: int
    timestamp: str
    data_item_id: str
    category: Literal["SAMPLE", "EVENT", "CONDITION"]
    type: str
    sub_type: str | None
    value_num: float | None
    value_str: str | None
    condition_level: Literal["NORMAL", "WARNING", "FAULT", "UNAVAILABLE"] | None
    condition_native_code: str | None
    condition_severity: str | None
    condition_qualifier: str | None


@dataclass(frozen=True)
class StreamsResult:
    instance_id: str
    first_sequence: int
    last_sequence: int
    next_sequence: int
    schema_version: str
    creation_time: str
    observations: list[Observation]


def _local(tag: str) -> str:
    return tag.split("}", 1)[-1] if "}" in tag else tag


def _parse_float_or_none(s: str | None) -> float | None:
    if s is None or s == "UNAVAILABLE":
        return None
    try:
        return float(s)
    except ValueError:
        return None


def parse_streams(xml: str) -> StreamsResult:
    root = etree.fromstring(xml.encode("utf-8"))
    header = next(c for c in root if _local(c.tag) == "Header")
    streams = next((c for c in root if _local(c.tag) == "Streams"), None)

    observations: list[Observation] = []
    if streams is not None:
        for dev_stream in streams:
            if _local(dev_stream.tag) != "DeviceStream":
                continue
            device_uuid = dev_stream.get("uuid", "")
            for comp_stream in dev_stream:
                if _local(comp_stream.tag) != "ComponentStream":
                    continue
                for category_node in comp_stream:
                    tag = _local(category_node.tag)
                    if tag not in CATEGORY_PARENTS:
                        continue
                    category = CATEGORY_PARENTS[tag]
                    for item in category_node:
                        itag = _local(item.tag)
                        if category == "CONDITION":
                            if itag not in CONDITION_LEVELS:
                                continue
                            observations.append(
                                Observation(
                                    device_uuid=device_uuid,
                                    sequence=int(item.get("sequence", "0")),
                                    timestamp=item.get("timestamp", ""),
                                    data_item_id=item.get("dataItemId", ""),
                                    category="CONDITION",
                                    type=item.get("type", ""),
                                    sub_type=item.get("subType"),
                                    value_num=None,
                                    value_str=item.text,
                                    condition_level=itag.upper(),  # type: ignore[arg-type]
                                    condition_native_code=item.get("nativeCode"),
                                    condition_severity=item.get("nativeSeverity"),
                                    condition_qualifier=item.get("qualifier"),
                                )
                            )
                        else:
                            text = item.text
                            observations.append(
                                Observation(
                                    device_uuid=device_uuid,
                                    sequence=int(item.get("sequence", "0")),
                                    timestamp=item.get("timestamp", ""),
                                    data_item_id=item.get("dataItemId", ""),
                                    category=category,
                                    type=itag,
                                    sub_type=item.get("subType"),
                                    value_num=_parse_float_or_none(text),
                                    value_str=text,
                                    condition_level=None,
                                    condition_native_code=None,
                                    condition_severity=None,
                                    condition_qualifier=None,
                                )
                            )

    return StreamsResult(
        instance_id=header.get("instanceId", ""),
        first_sequence=int(header.get("firstSequence", "0")),
        last_sequence=int(header.get("lastSequence", "0")),
        next_sequence=int(header.get("nextSequence", "0")),
        schema_version=header.get("schemaVersion", header.get("version", "")),
        creation_time=header.get("creationTime", ""),
        observations=observations,
    )

File: cloud/src/xml/streams.ts
import { XMLParser } from "fast-xml-parser";

export type ParsedObservation = {
  deviceUuid: string;
  sequence: number;
  timestamp: string;
  dataItemId: string;
  category: "SAMPLE" | "EVENT" | "CONDITION";
  type: string;
  subType?: string;
  valueNum: number | null;
  valueStr: string | null;
  conditionLevel?: "NORMAL" | "WARNING" | "FAULT" | "UNAVAILABLE";
  conditionNativeCode?: string;
  conditionSeverity?: string;
  conditionQualifier?: string;
};

export type StreamsParseResult = {
  header: {
    instanceId: string;
    firstSequence: number;
    lastSequence: number;
    nextSequence: number;
    schemaVersion: string;
    creationTime: string;
  };
  observations: ParsedObservation[];
};

const parser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: "@_",
  parseAttributeValue: false,
  allowBooleanAttributes: true,
  preserveOrder: false,
  textNodeName: "#text",
  isArray: (tagName) => {
    // every known observation-bearing tag should be an array
    if (["DeviceStream", "ComponentStream"].includes(tagName)) return true;
    return false;
  },
});

// categories map from XML parent container name to MTConnect category
const CATEGORY_PARENTS: Record<string, "SAMPLE" | "EVENT" | "CONDITION"> = {
  Samples: "SAMPLE",
  Events: "EVENT",
  Condition: "CONDITION",
};

const CONDITION_LEVEL_TAGS = ["Normal", "Warning", "Fault", "Unavailable"];

export function parseStreams(xml: string): StreamsParseResult {
  const root = parser.parse(xml);
  const ms = root.MTConnectStreams;
  const hdr = ms.Header;

  const observations: ParsedObservation[] = [];

  const devStreams = ms.Streams?.DeviceStream ?? [];
  const devArr = Array.isArray(devStreams) ? devStreams : [devStreams];

  for (const ds of devArr) {
    const deviceUuid = ds["@_uuid"];
    const compStreams = ds.ComponentStream ?? [];
    const compArr = Array.isArray(compStreams) ? compStreams : [compStreams];
    for (const cs of compArr) {
      for (const [parentTag, inner] of Object.entries(cs)) {
        if (!(parentTag in CATEGORY_PARENTS)) continue;
        const category = CATEGORY_PARENTS[parentTag];
        collectFromCategoryNode(deviceUuid, category, inner, observations);
      }
    }
  }

  return {
    header: {
      instanceId: hdr["@_instanceId"],
      firstSequence: Number(hdr["@_firstSequence"]),
      lastSequence: Number(hdr["@_lastSequence"]),
      nextSequence: Number(hdr["@_nextSequence"]),
      schemaVersion: hdr["@_schemaVersion"] ?? hdr["@_version"] ?? "",
      creationTime: hdr["@_creationTime"],
    },
    observations,
  };
}

function collectFromCategoryNode(
  deviceUuid: string,
  category: "SAMPLE" | "EVENT" | "CONDITION",
  node: unknown,
  out: ParsedObservation[],
): void {
  if (node === null || node === undefined) return;
  // node is { TagName: [items] or item, ... }
  for (const [tag, value] of Object.entries(node as Record<string, unknown>)) {
    const arr = Array.isArray(value) ? value : [value];
    for (const item of arr) {
      if (!item || typeof item !== "object") continue;
      const it = item as Record<string, string>;
      if (category === "CONDITION") {
        // tag is the level (Normal|Warning|Fault|Unavailable)
        if (!CONDITION_LEVEL_TAGS.includes(tag)) continue;
        out.push({
          deviceUuid,
          sequence: Number(it["@_sequence"]),
          timestamp: it["@_timestamp"],
          dataItemId: it["@_dataItemId"],
          category: "CONDITION",
          type: it["@_type"] ?? "",
          subType: it["@_subType"],
          valueNum: null,
          valueStr: (it as Record<string, string>)["#text"] ?? null,
          conditionLevel: tag.toUpperCase() as
            | "NORMAL"
            | "WARNING"
            | "FAULT"
            | "UNAVAILABLE",
          conditionNativeCode: it["@_nativeCode"],
          conditionSeverity: it["@_nativeSeverity"],
          conditionQualifier: it["@_qualifier"],
        });
      } else {
        const text = (it as Record<string, string>)["#text"];
        const valueStr = text ?? (typeof item === "string" ? item : null);
        const valueNum =
          valueStr !== null && valueStr !== "UNAVAILABLE"
            ? parseFloatOrNull(valueStr)
            : null;
        out.push({
          deviceUuid,
          sequence: Number(it["@_sequence"]),
          timestamp: it["@_timestamp"],
          dataItemId: it["@_dataItemId"],
          category,
          type: tag,
          subType: it["@_subType"],
          valueNum,
          valueStr,
        });
      }
    }
  }
}

function parseFloatOrNull(s: string): number | null {
  const n = Number(s);
  return isNaN(n) ? null : n;
}
